runtime/vm/resolvesupport.cpp

```diff
@@ -1851,6 +1851,122 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
 
 }
 
+j9object_t
+resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags)
+{
+	J9RAMConstantDynamicRef *ramCPEntry = (J9RAMConstantDynamicRef*)ramCP + cpIndex;
+	j9object_t value = NULL;
+	J9JavaVM *vm = vmThread->javaVM;
+	bool resolved = false;
+
+retry:
+	/* Fast path resolution for previously resolved entries */
+	if (NULL != ramCPEntry->value) {
+		value = ramCPEntry->value;
+		resolved = true;
+	} else {
+		/* check if already resolved to NULL or exception */
+		if (NULL != ramCPEntry->exception) {
+			J9Class *throwable = J9VMJAVALANGTHROWABLE_OR_NULL(vm);
+			J9Class *clazz = J9OBJECT_CLAZZ(vmThread, ramCPEntry->exception);
+			if (ramCPEntry->exception == vm->voidReflectClass->classObject) {
+				/* Resolved with null reference */
+				resolved = true;
+			} else if (isSameOrSuperClassOf(throwable, clazz)) {
+				/* Resolved with exception */
+				VM_VMHelpers::setExceptionPending(vmThread, ramCPEntry->exception);
+				resolved = true;
+			}
+		}
+	}
+
+	if (resolved) {
+		return value;
+	} else {
+		/* Slow path, attempt to acquire resolve mutex or retry, or wait */
+		omrthread_monitor_enter(vm->constantDynamicMutex);
+		if (NULL == ramCPEntry->value && NULL == ramCPEntry->exception) {
```

Comment:
After entering the monitor, you should do all of the checks from before the monitor enter again, like:

enter monitor
checks
if (!resolved) {
}
exit monitor
...fall through to return...

Reply:
inside the monitor it have a simplified block of checks that determine if there have been any update to the cpEntry and redirects with `goto retry;`

Reply:
I'm not convinced it's correct, and there's no point trying to optimize the slow path instead of keeping the code the same as outside for understandability.